lwgenerate/hcl.go

```diff
@@ -0,0 +1,366 @@
+func createMapTraversalTokens(input map[string]string) hclwrite.Tokens {
+	// Sort input
+	keys := []string{}
+	for k := range input {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	tokens := hclwrite.Tokens{
+		{Type: hclsyntax.TokenOBrace, Bytes: []byte("{"), SpacesBefore: 1},
+		{Type: hclsyntax.TokenNewline, Bytes: []byte("\n")},
+	}
+
+	for _, k := range keys {
+		tokens = append(tokens, []*hclwrite.Token{
+			{Type: hclsyntax.TokenStringLit, Bytes: []byte(k)},
+			{Type: hclsyntax.TokenEqual, Bytes: []byte("=")},
+			{Type: hclsyntax.TokenStringLit, Bytes: []byte(" " + input[k]), SpacesBefore: 1},
+			{Type: hclsyntax.TokenNewline, Bytes: []byte("\n")},
+		}...)
+	}
+
+	tokens = append(tokens, []*hclwrite.Token{
+		{Type: hclsyntax.TokenNewline},
+		{Type: hclsyntax.TokenCBrace, Bytes: []byte("}")},
+	}...)
+
+	return tokens
+}
+
+// Convert blocks to a string
+func CreateHclStringOutput(blocks []*hclwrite.Block) string {
+	file := hclwrite.NewEmptyFile()
+	body := file.Body()
+
+	for _, b := range blocks {
+		body.AppendBlock(b)
+		body.AppendNewline()
+	}
+	return string(file.Bytes())
+}
+
+// Create required providers block
+func CreateRequiredProviders(providers ...*HclRequiredProvider) (*hclwrite.Block, error) {
+	block, err := HclCreateGenericBlock("terraform", nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	providerDetails := map[string]interface{}{}
+	for _, provider := range providers {
+		details := map[string]interface{}{}
+		if provider.Source() != "" {
+			details["source"] = provider.Source()
+		}
+		if provider.Version() != "" {
+			details["version"] = provider.Version()
+		}
+		providerDetails[provider.Name()] = details
+	}
+
+	requiredProviders, err := HclCreateGenericBlock("required_providers", nil, providerDetails)
+	if err != nil {
+		return nil, err
+	}
+	block.Body().AppendBlock(requiredProviders)
+
+	return block, nil
+}
+
+// helper to create a hcl.Traversal in the order of supplied []string
+//
+// e.g. []string{"a", "b", "c"} as input results in traversal having value a.b.c
+func CreateSimpleTraversal(input []string) hcl.Traversal {
+	traversers := []hcl.Traverser{}
+
+	for i, val := range input {
+		if i == 0 {
+			traversers = append(traversers, hcl.TraverseRoot{Name: val})
+		} else {
+			traversers = append(traversers, hcl.TraverseAttr{Name: val})
+		}
+	}
+	return traversers
+}
+
+// Simple helper to combine multiple blocks (or slices of blocks) into a single slice to be rendered to string
+func CombineHclBlocks(results ...interface{}) ([]*hclwrite.Block, error) {
+	blocks := []*hclwrite.Block{}
+	// Combine all blocks into single flat slice
+	for _, result := range results {
+		switch v := result.(type) {
+		case *hclwrite.Block:
+			if v != nil {
+				blocks = append(blocks, v)
+			}
+		case []*hclwrite.Block:
+			if len(v) > 0 {
+				blocks = append(blocks, v...)
+			}
+		default:
+			return nil, errors.New("Unknown type supplied!")
+		}
+	}
+
+	return blocks, nil
+}
```

Comment:
As much as I like the fact that you are helping the user combine blocks, I don't think we should throw an error here, but instead, only add the two cases you have for a single pointer and an array of pointers. 

```suggestion
func CombineHclBlocks(results ...interface{}) []*hclwrite.Block {
	blocks := []*hclwrite.Block{}
	// Combine all blocks into single flat slice
	for _, result := range results {
		switch v := result.(type) {
		case *hclwrite.Block:
			if v != nil {
				blocks = append(blocks, v)
			}
		case []*hclwrite.Block:
			if len(v) > 0 {
				blocks = append(blocks, v...)
			}
		default:
			continue
		}
	}

	return blocks
}
```

Reply:
👍